### ISSUE TITLE
Applications: Some POST requests are replied to with code 501

### ISSUE THREAD

**Issue by apiening** [gravitational/teleport#19498]
Expected behavior:
Accessing a web portal with `teleport` `Applications` should behave the same way like accessing the target web portal directly. Specifically: All requests should pass with the same results.

Current behavior:
Some **POST** requests (**XHR**) are getting **501** responses. This breaks the functionality of an **Application** when some functions are not working.

Bug details:
- Teleport version: `11.1.3`
- Recreation steps: I have this issue specifically when I access the web interfae of a `Proxmox VE` instance with a **teleport** **application**: I can log in and do all sorts of things but when I try to open a **shell** (**xterm.js**) the **POST** request fails with status **501**. Same is true for the **Start** and **Shutdown** actions for a selected **VM**. When I access the web interface directly or over an **SSH tunnel**, everything is working fine.
- Debug logs:
```
INFO             Round trip: POST /api2/json/nodes/pve1/termproxy, code: 501, duration: 1.015604ms tls:version: 304, tls:resume:false, tls:csuite:1301, tls:server:<hash>.teleport.cluster.local forward/fwd.go:182
```

I know that this issue is very specific, because it only happens on some requests on a specific application. But it happens on all instances running different versions of **PVE**.
The strange thing is, that it was working for my for quite a while: I then updated from `teleport 10` to `teleport 11` and did a **cert rotation** and since then it stopped working.

I did also open a thread in the **Proxmox** forum on this topic: https://forum.proxmox.com/threads/reverse-proxy-start-stop-vms-and-xterm-js-not-working.119695/

**Comment by AlexRRB** [comment]
i think, it is because Teleport doesn't support websockets for webapps ((
it will work after they add this functional

**Comment by apiening** [comment]
> i think, it is because Teleport doesn't support websockets for webapps (( it will work after they add this functional

I'm not sure if the web app really uses websockets, at least not for the **Start** and **Shutdown** operations. When I click on the **Shutdown** button, I can see a request with the type **XHR** in the browsers **DevTools**. It has the option `:method: POST` in the request header.
The response to this request with the url `https://myapp.teleport.mydomain.com/api2/extjs/nodes/pve1/qemu/101/status/shutdown` looks interesting to me:
```json
{"data":null,"status":501,"message":"Method 'GET /nodes/pve1/qemu/101/status/shutdown' not implemented","success":0}
```
So the browser makes a **POST** (**xhr**) request to the **teleport** application, and **teleport** makes a **GET** response to the backend app instead?
Do I get this right?
What may cause this?

**Comment by justinforlenza** [comment]
Im running into the same issue with my Proxmox VE application service. 

However I am not encountering the same log message. In my Proxmox VE access log it does show a POST to the API endpoint:

```
- - [10/01/2023:21:45:10 -0500] "POST /api2/extjs/nodes/33K891HPY01/lxc/600/status/start HTTP/1.1" 501 -
```
But still a 501

**Comment by apiening** [comment]
My log message was extracted from the **teleport** service log. I have searched for the URI parameters and found this entry, but I cannot rule out that this was not a request by the web interface. I may have caused this **GET** request by myself playing around with the browsers web developer tools.

I'm still suffering from this issue. Searching for similar symptoms I found reports with status **501** errors happening when HTTPS upgrade fails. The solution often involved adding special header values on the proxy side.

However, since the PVE web interface is served via HTTPS already and the teleport application proxy is a HTTPS endpoint as well, I'm not sure if this even applies. And it would probably break the web interface all together and not only these specific calls.
But I'm guessing here, I do not understand this topic well enough to track the cause of this issue.

**Comment by justinforlenza** [comment]
Did some digging looks like someone in the discussions (#6746) was able to resolve it by added a nginx proxy between teleport and proxmox.

I replicated that on my instance, and it also resolved it for me.

Heres my nginx config for reference.
```
server {
    listen 127.0.0.1:443;
    server_name _;
    ssl on;
    ssl_certificate /etc/pve/local/pve-ssl.pem;
    ssl_certificate_key /etc/pve/local/pve-ssl.key;
    proxy_redirect off;
    location / {
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade"; 
        proxy_pass https://localhost:8006;
	proxy_buffering off;
	client_max_body_size 0;
	proxy_connect_timeout  3600s;
        proxy_read_timeout  3600s;
        proxy_send_timeout  3600s;
        send_timeout  3600s;
    }
}
```
So i guess it might be some of the proxy settings being set in nginx, that cant be set in the teleport proxy configuration?

**Comment by apiening** [comment]
> So i guess it might be some of the proxy settings being set in nginx, that cant be set in the teleport proxy configuration?

I agree that this kind of proves that the issue has to be related with the teleport proxy behaviour.

Looking at the documentation for Application Access, there is not so much that could be changed other than some rewrite options (see https://goteleport.com/docs/application-access/reference/).

I can't find anything special in the `nginx config`, other than these two header values: `Upgrade` and `Connection`. I'll try to experiment with these options with the rewrite: / headers: option from the `teleport` apps config.

The timeout settings are set to 1 hour (3600s), but I don't believe that this is the cause of the issue since it occurs immediately and not after a while when a timeout may run out.

**Comment by apiening** [comment]
I think this issue is related to https://github.com/gravitational/teleport/issues/18636.
But it looks like it could not be solved by setting static header values, since they need to be set dynamically based on the client headers.

**Comment by apiening** [comment]
@justinforlenza I have tried to setup a reverse proxy with `nginx` based on your configuration and it does work for me this way as well.

Can someone of the developers please take a look at this?
This error is reproducible and even though there seems to be a workaround with an additional proxy server, this is quite a limitation of the `application-access` feature.

If I can provide additional information or if I should try a test version, please let me know.
This issue is bugging me, so I would love to see this addressed.

**Comment by apiening** [comment]
Upgrading teleport to version `12.0.2` seems to solve this issue for me.
Can someone confirm?

**Comment by aeliusrs** [comment]
Hey all, I have the same exact issue with Teleport 13.3.4 and PVE 8+

Is there any solution outside of setting up the nginx reverse proxy ? we use traefik in my company...

**Comment by marcusaram** [comment]
Same issue here, I'm not sure but can't remember that this occurred before upgrading Teleport.

**Comment by rezarajan** [comment]
> Same issue here, I'm not sure but can't remember that this occurred before upgrading Teleport. I can open a console on Proxmox but can't start a VM.

Same issue here - with Teleport 13.2.3 & PVE 7.4-16. Console works just fine, but doing just about anything else with VMs results in a 501.

**Comment by rrizun** [comment]
same issue here .. "Connection failed (Error 501: Not Implemented)" when starting proxmox's termproxy (e.g., node shell or container shell) .. teleport 13.3.6 and pve 7.4

**Comment by R-Nabil** [comment]
Same issue here

**Comment by jm-bertheas-datailor** [comment]
Same issue here - with Teleport 13.3.5 & PVE 8.0

**Comment by jm-bertheas-datailor** [comment]
Teleport upgrade from 13.3.5 to 13.3.8 has no effect.

**Comment by R-Nabil** [comment]
> Teleport upgrade from 13.3.5 to 13.3.8 has no effect.

Same on my side. Upgrade didn't help

**Comment by R-Nabil** [comment]
On my side, after downgrading to 13.3.1 it worked. It seems it start breaking after 13.3.3

**Comment by rrizun** [comment]
nice find same here .. 13.3.1 works, 13.3.3 broken

**Comment by R-Nabil** [comment]
In case anyone's interested, 13.4.0 is also broken, no point upgrading for a fix...

**Comment by webvictim** [comment]
To everyone reporting this; can you share your Teleport `app_service` config here?

I suspect that this is a CORS issue where Proxmox is refusing to honour requests coming from Teleport. The 501 "Not Implemented" response is very odd as usually it's a straight 403 "Forbidden" instead in these scenarios.

To try and prove this, you can manually overwrite the `Host` and `Origin` headers in your `app_service` config to the Teleport-proxied versions:

```yaml
app_service:
  enabled: true
  apps:
  - name: proxmox
    uri: http://localhost:8006
    public_addr: proxmox.teleport.example.com
    rewrite:
      headers:
      - "Host: proxmox.teleport.example.com"
      - "Origin: https://proxmox.teleport.teleportdemo.com"
```

This may be a dead end, but it'd be worth knowing that it's something different to this.

**Comment by R-Nabil** [comment]
Hi,

Here's my app_service :

`    - name: "proxmox"
      uri: "https://proxmox.home.mydomain.com"
      labels:
        environment: admin
`
Not sure I understand the difference between :

proxmox.teleport.example.com and https://proxmox.teleport.teleportdemo.com

**Comment by marcusaram** [comment]
Rewriting the headers doesn't work, we have the same issue, just 501 not implemented. We do run multiple agents to connect to PVE. And the agents do connect to multiple servers, but I just tried to run 1 instance against 1 PVE host. But that doesn't make a difference

I will setup a test/lab environment and use mitm-proxy to debug what kinds of requests are going to pve.

**Comment by R-Nabil** [comment]
Happ to help with any type of investigation or test i could do. I'm in a lab envt.

**Comment by EternalFr0st** [comment]
Just FYI - as a workaround, you can use the TCP App Access instead: https://goteleport.com/docs/application-access/guides/tcp/

It's not as convenient as the general application access functionality, but doesn't suffer from the same problems.
Tested with Teleport 13.3.8 and PVE 8.0.4

**Comment by arsenaultja** [comment]
FYI, 13.4.0 seems to have fixed it for me. Can anyone back this up or am I just crazy?

**Comment by R-Nabil** [comment]
I think on my side I'm more likely to give up on using teleport. I do not understand that a feature can be fixed/broken multiple times between versions and it makes it a headache to have to manage it.
It's a shame the product really is interesting apart from that

**Comment by DocterDum** [comment]
Updating to 14.0.0 fixed this for me :)

**Comment by R-Nabil** [comment]
> Updating to 14.0.0 fixed this for me :)

Would you mind sharing your app_service ?

**Comment by DocterDum** [comment]
> > Updating to 14.0.0 fixed this for me :)
> 
> Would you mind sharing your app_service ?
It's pretty simple - 
app_service:
  enabled: "yes"
  debug_app: false
  apps:
  - name: my-app-name
    uri: https://1.2.3.4:8006/
    public_addr: ""
    insecure_skip_verify: true

**Comment by marcusaram** [comment]
Can confirm that Teleport v14 fixes this issue. I was working on a workaround that uses another reverse proxy between teleport and Proxmox PVE. Glad to see that the issue is fixed.

**Comment by webvictim** [comment]
I suspect this is as a result of https://github.com/gravitational/teleport/pull/27761 being merged (which came out in 14.0.0)

**Comment by aeliusrs** [comment]
I confirm that the update to version 14 Fix the issues, even without a third-parties load balancing :tada:
